### ISSUE TITLE
Stricter types for `JSON`

### ISSUE THREAD

**Issue by brunocrosier** [microsoft/TypeScript#51003]
# lib Update Request

## Configuration Check

My compilation *target* is `ES2015` and my *lib* is `the default`.

## Missing / Incorrect Definition

`JSON.parse()` and `JSON.stringify()`

The definitions for both of these functions currently accept a `string`, but this is not strict enough. Not all strings can be parsed or stringified. Many of these can be caught at a type level, meaning they could and should be added to the type definitions.

## Sample Code

```ts
// @ts-expect-error
JSON.parse('')

// @ts-expect-error
JSON.parse("'foo'");

// @ts-expect-error
JSON.parse("[1, 2, 3, 4, ]");

// @ts-expect-error
JSON.parse('?')

// @ts-expect-error
JSON.parse("bar")

// @ts-expect-error
JSON.stringify({
    key: 2n
})

const circularReference: { myself?: any } = {};
circularReference.myself = circularReference;

// @ts-expect-error
JSON.stringify(circularReference)
```

TS Playground: https://www.typescriptlang.org/play?target=99#code/PTAEAEBcGcFoFMAeAHeBjSCBOWD2WAoAKQGUB5AOQDpkBDLaeACgHIWBKAgkCGBFdJng58xctToNmAIhYAzXLhbT2Abi48ocJKgzY8hUpRr1GTaQG0AjABpQAJjsBmOwBY7AXRXruYLf10hEUNxEylWAH4ODT8+HUF9USMJUxkAI3oVGN5tAT1hAzFjaEgsAEsAOwBzMrkATyYAbwJQVtAAa3g6gC4HCoIAX04CNFwKktA0Mqw0AFcAG3oAJXg5YXgKtHhextAAWzrGebkI3toKutAB0ABeUEaB9SmZheXV9c34KgOjuVvJ6ZzRZYFZrLAbLY+TRxPJBQrJKglcrVWoNZ5At5giHwThAA
## Documentation Link

[JSON](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON)
[JSON.parse](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/parse)
[JSON.stringify](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify)

**Comment by fatcerberus** [comment]
This doesn’t seem very useful; it only works if you’re passing string literals to ‘JSON.parse()`, so the improved type checking is purely academic.  Also, how do you propose to detect circular references at type level without excluding valid recursive structures?

**Comment by brunocrosier** [comment]
> This doesn’t seem very useful; it only works if you’re passing string literals to `JSON.parse()`—it’s not possible to make a “JSON-compatible string” type without relying on generic inference—so the improved type checking is purely academic. Also, how do you propose to detect circular references at type level without excluding valid recursive structures?

You are mistaken. 

https://www.typescriptlang.org/play?target=7#code/C4TwDgpgBAwglgJwMYFcA2BDBAlCAzCBCAOyQgB4AVAPigF4oBvAKCjagG0BrCEALigBnYAjjEA5lAA+UYigC2AI0LShIJQHs0AXQGVmAX2ahIUACJxBGNGg0B3CABMAUgGUA8gDkAClkERKCAAPYHooAHJw1QADABJGYVEJA3jwgBptKJl0gyyItKhM1XSoXINo43BoCysbeyc3LwA1axQKGjDFOHExUJkWdk4efiERMUkZOSUVGUF1RS1dKC6e4mBDVXhkdCxcAiJSdupmStMAQTqHFw8fPwDg0IZwkTa88Lxrfze5Gzy4ximygQ5RiACJ4olxgZQdEYhwIWNktpYTJoowEUlxAZyideoQPmQoI1PAB9dwANyIokc0AG7AA9AAqRmsQaM2AaYjkwjAQRQDBEjDkjCuJCiMChdyKABWECQoU8GmAGGAcE5UAAFMSAJSjTFQXoafnEKAaGVy4AAOlZ7HZAAEwFgMPIoMAHlAzlBhWg4I4iTc9eNrYMoPbHQhnVAiOS4NyEB6oHgUKRVergAALFWuiPEQR4DQIeR8jPQIiCdC8y1QSjpyyJ5PytUmutIaxoJyJgtQCAYJDpqDyCDTeMaPCu9PQM2y+XBtlQACSY4Fg+HUCQnOVYj5xAgwg7U4tggKJdku7dfoP8r5WGgIgwufzhY7ykft4nUHDJFCl9CllntvpG0PzuKhuxCEhHD5SEJGoDU3RCPQwLdYhIPMSw23qa4vF8BB-ECEIoAAflPOMoEQh4IL5C5bCuYkcLw91iMoMiSMIApo1jQhCIEODa0EAR7xAAphgEaDxAKb02gE4gQF1OhaEE7VpJAABuICmRZEN2RgTk415flBWFUVxVCSTbyNAVnCFEUxTgCUoClacFSVFUm01HVAwkf82DDJ0XTMhMrKM2z7LMgoUHLNsQGNU1zXlU14ywCMhNdI1lDXXSeScbzQygB0-KjCAwEwMh409JMUzcjMszvB8CyLcdS13CtBBy3yIxdQRHUJM5HFQsQaTWVzOQKOxazdIRuogAp7z9H0d2WIgMC4NdMwjeVCGLI0TyIYAUAQYgAFoAuJV13WAI15GW6A4FCHtBDgFQLsKjBHDawDBjEuA8BAKhYLM8jwJQvkagw2ibhaNA2j+ojWPjPR2KKkquJ4jNLGU4TeFExFxK9VoIGUuSFJkgout7CBuM8iZZAUIElKptTBg0oDcp0rkeWvQybJMvGofMgygu5uzJTi5zlVTE0tRuXUxLavLw0jALPUF4zhd5tpwsimxovvWKnIS-kEGS480ugdd2YQc85fyjrCuK8mypNJKMGi0cqevFCaeHYtM1CXt9IwD3+TAMAEA0bk5ssUJHyECB20bCRGr1i0PzDyBLcen2szsOAbGWaAvrwR63pZ9rIzJnq+r5Aav2G4hRvGguppmz35ugRQlpWvsnQ23DUqT3b9qOk6A3g0JnqungDTuwPHvjZ6lpLrSPvYQvfpoDUAerJDKLQ2oaIaCH8Zh4id1IhG7eRhBKY1QEZip7UOG0VQfjQUmpspsSX9pwh6bEtSjDMBpEgTARAMq5lCMSMklJCC+gJv6Lw0CqRwJOMwek9I8q8kOsESA8psFGwLMwKBFJkE0ktOGfwGpIjahOOgzBghsFBFwcAfBYcEBEJuEg2BZCKEQA1KCd4GgNDhFBNqNSaCMF2iwTgi0rDCHEJgdSCA5C7j8I4AARgKAAJgKAAZgKAAFgKNoUR4i6FSIYTIvBhA2EcMQSQ7hyjeFUMIuEGhEj6GMOYXI9hCjSFONUaCRQWBRG0MkdIphsibHyM4Q4pRlo14ajpGwESUAtHEEMO482wg1yIFQKAvYhASBkAEIwAcIB-BoDwJTQSpQwiMAMGpJAeSdg4HwEUw4lp5AVLjmOBgzTtgFPaQcMgZjwmWMidYghvjYmKLgQknG30QAagGfk3YwzikQG1EAA

**Comment by fatcerberus** [comment]
That would exclude a ton of perfectly JSON-able objects.

**Comment by brunocrosier** [comment]
> That would exclude a ton of perfectly JSON-able objects.

For example?

**Comment by brunocrosier** [comment]
@fatcerberus can you provide an example of something which this PR does not correctly parse or stringify?

Here you will find examples of a [tree structure being parsed and stringified correctly](https://www.typescriptlang.org/play?target=7&ssl=65&ssc=1&pln=65&pc=14#code/C4TwDgpgBAwglgJwMYFcA2BDBAlCAzCBCAOyQgB4AVAPigF4oBvAKCjagG0BrCEALigBnYAjjEA5lAA+UYigC2AI0LShIJQHs0AXQGVmAX2ahIUACJxBGNGg0B3CABMAUgGUA8gDkAClkERKCAAPYHooAHJw1QADABJGYVEJA3jwgBptKJl0gyyItKhM1XSoXINo43BoCysbeyc3LwA1axQKGjDFOHExUJkWdk4efiERMUkZOSUVGUF1RS1dKC6e4mBDVXhkdCxcAiJSdupmStMAQTqHFw8fPwDg0IZwkTa88Lxrfze5Gzy4ximygQ5RiACJ4olxgZQdEYhwIWNktpYTJoowEUlxAZyideoQPmQoI1PAB9dwAN0Iokc0AG7AA9AAqRmsQaM2AaYiUhDAQRQDBEjDkjCuJCiMChdyKABWECQoU8GmAGGAcE5UAAFMSAJSjTFQXoafnEKAaGVy4AAOlZ7HZAAEwFgMPIoMAHlAzlBhWg4I4iTc9eNrYMoPbHQhnVAiOS4NyPVA8ChSKr1cAABYq10R4iCPAaBDyPnp6BEQToXmWqCUNOWBNJ+Vqk21pDWNBOBP5qAQDBINNQeQQaYIU14V1p6Bm2Xy4NsqAASVHAoHQ6gSE5yrEfOIEGE7cnFsEBWLsh3br9+-lfKw0BEGBzeYL7eUD5v46g4ZIoQvoUsM9t9Jtd87ioLsQhIRw+UhCRqA1N0Qj0UC3WICDzEsVt6muLxfAQfxAhCKAAH4TzjBCHnAvkLlsK5iWw3D3SIygoAEbduQKaNY0IAiBFgmtBAEO8QAKYYBCg8QCm9Np+OIEBdToWgBO1KSQAAbkApkWRDdkYE5bleX5QVhVFcVQgkm8jQFZwhRFMU4AlKApSnBUlRVRtNR1QMJD-NgwydF1TPjSzDJsuzTIKFAy1bEBjVNc15VNYcsAjQTXSNZRVx0wgzy80MoAdXyowgMBMDIYdPUTZNXPTTNb3vfNCzHEsd3LQRsp8iMXUER1CTORwULEGk1hczkCjsGs3SELqIAKO8-R9bdliIDAuFXDMI3lQgiyNY8iGAFAEGIABafziVdd1gCNeQlugOBQm7QQ4BUc6CowRxWoAwZRLgPAQCoGDTNIsDkL5Gp0Oom4WjQNpfsI4iVD0NjCuKzjuPTSwlKE3gRMRMSvVaCAlNk+TpIKTqewgLiPImWQFCBRTKdUwZ1MAnLtK5TKrwM6zjNxyGzP0wKudsyVYqc5UUxNLUbl1UTWty8NI38z0BaMoWebaMKIpsKK7xixz4v5BAkqPVLoDXNmeScWW8vagqirJ0qTUSjAoo0UdRKvZDqaHIsM1CHs9IwD3+TAMAEA0SlZssUIHyECA2wbCQGt1i13zDyAeQen3MzsOAbGWaBPrwB7XuZtrI1J7rer5frPyG4gRrGgvJumz25ugRRFuW3snXWnCUqTna9sO46Azg0InsungDVuwOHuHJ7FpLzT3vYQufpoDV-qrRDyNQ2oqIacG8ehoiWLhqsEbtkqKY1QEZkp7UOG0VQfjQEnJop0SX5pwg6dE1SjDMBpEgTARB0o5lCMSMk3JqT439F4aBVJfQQBOFAikSCaSWjXhqbUJx6T0lyryA6wRIDymIYbfMzA0EwOQZacM-gNSRFwcwfBhDBDEKCKQ4A5Cw4ICoTcRBhtaH0IgBqUE7wNAaHCKCbUqkWEELtEQkhFoeGUOoRgiAdC7hiI4AARgKAAJgKAAZgKAAFgKNoGRcjWGKPYcoshVI1ECPQUIzBIjGEEXCMw2xSjOEqKcXw9RbjNEeNBIoLAMi8EKL8Vw1RQSXE0MwdgukbBhJQAMcQQwzCzbCFXIgVAoC9iEBIGQAQjB+wgH8GgPAFMBKlDCIwAwqkkAFJ2DgfAJTDiWnkFUuOo4GCtO2EUzpBwyA2JifY-xjiKEJIQa42BWDsZfRABqIZhTdijNKRAZhwTFkpMAmuIg5TmbsEcCqDAAgOCnJDEwG5tz2BjwEKCE6ioaRQF0aCNI9yHmjBVHA1JvygUaEgNuRwAgXhTR+Q8gw3ygUPN7LnRwBwrnQvhYC+F8KnlQBeQGN50BdGWk+XCzFvzYVoqBRi0lDzsW4q8FAfFHzLQGK+RSslJLqUhipZywYtLXkaHeYS4xrKeUhnJaKwY3KJV8rxQKgllozEiolWwcVyq7lqpDDK+ljLCUAFYlXKtVZy7QHKgVGspWyx5Dxnn8veSy01pLhD-JORqkFJAnAQoQOrS1KqHWYsRWgZFJBUUaqlaKrVngGVyoyUSg1ErzWirDTyiNUa7XMrjaKhNPKk2cpTYygxlphV+s5VmzlObqV5ujQWxVxbqWlupeW0lla036traS+tvyTUUvrV2wY5KDCyNQYkjRWicKiPCIwUERyICgj4JO85ypZ0cEnWPWddLI06q+aCJ1bpZ2TrdWC2dkLYVTprIGg4S6V0PDXba+VnzYVXpCDe2Vgr00PtBKuvg67U3yuFe+z936dUKtBP+69X7b1Mv1QYbQoGn3gZfdAe126xYzrnaCA9Tgj1eogCegNQbiCXo-WBwDVbY2weAM+7VpGWXkcoxu0jf60iPoo-BqjabFW0dY-RltIGYMwdyNqIAA):

**Comment by fatcerberus** [comment]
My point was, your types are excluding `.myself` on the basis that it *could* be a self-reference.  That’s way too strict.  As for `JSON.parse`, those stop being useful the minute you start passing string variables to it instead of hard-coded strings.

**Comment by brunocrosier** [comment]
> My point was, your types are excluding `.myself` on the basis that it _could_ be a self-reference. That’s way too strict. As for `JSON.parse`, those stop being useful the minute you start passing string variables to it instead of hard-coded strings.

```ts
const emptyString = ''
JSON.parse(emptyString) // no error
```

![image](https://user-images.githubusercontent.com/18399089/193172543-a741fe99-eead-4a28-92ba-9a837abf71f8.png)

**Comment by fatcerberus** [comment]
Cases in point:
```ts
const circularReference: { myself?: any } = {};
circularReference.myself = { foo: 812 };
JSON_Override.stringify(circularReference);  // error?!
```

```ts
const str = '{ "foo": "validJSON" }';  // pretend this was read from a file...
JSON_Override.parse(str);  // error. this is not useful.
```

The types only exist at compile time.  It seems like you might not fully understand how the type system works?

**Comment by fatcerberus** [comment]
```ts
let str = '{ "foo": "validJSON" }';
JSON_Override.parse(str);  // error
```

**Comment by brunocrosier** [comment]
> ```ts
> let str = '{ "foo": "validJSON" }';
> JSON_Override.parse(str);  // error
> ```
> 
> Like I said, these extra checks only work if you're passing literal-typed strings to `parse`.

you're just pasting code and adding `// error` 

[here is a TS Playground ](https://www.typescriptlang.org/play?#code/C4TwDgpgBAIglgZwIYBsUHsDuEAmApAZQHkA5ABSQCcEIAVCAD2CgF4oBydqAHygAMAJAG8EwSnAB2AcwC+w9gBoAul16KZqjgqgqeWqBpl8AUKEixEqDNnzESANVQBXaGwBGcKZOa8hxqAFQANoA1hAgAFxQouLSehJOALZuEJR6CCDJ6ChKUR5eEsDGMsam4NAAgmhYuISkFNR0jMxs7GIumuwAZqg0nQlomoJCCcmpRnp8AETCMZKyU3yTQbNi8zJKS7x8Qquxskal3qk9AMbQdSQA+kQAbqniONB+gVAA9ABUH-6vH1AAwugJPdKMAEFAkFA8EhbkgCKdxGBmEQ3AArCCnZgkdDAJDAOBAqAACkuAEpoms4t50BCJFB0GiMcAAHQ-QJ-AACYCoSESUGAzSgFSgsJQcBwULsFP2rNeUE53MovKglAgtzgIKFUC6TgkmIJdOAAAs8fylRIEF10JREuDjdBVQgnCgwcyoLQjYhtbr9YSvacrLhtdaoBAkKcjVBEhAxml0F1+UboAz0ZjZb8oABJBOQ6OxqCnIG4yTgiQQURBlNMhDae1QMsViVVzHgqjQMRIC1Wm1BlLd9tJqCKiCFemMzFQRDp9lvNlDqg0AA8tFDTBHOHBc2kAD4iQKmFEV811+D4MhqjZLg0aPQmFAAPz1tWpKCH1cCiQboUX2p2a9NO9HxXKIyxBbRVXVEF7yiPdPQQKJOxAbQwkiaV5m0UUXAQiQQHJFht1pXDsJAABuOdPm+OU-kBYFUjBCEoRhOEETgJERWcdsaUhaFYXhRFkXHLEcTxA1iTJNDpGnAIFR5PlMMqRjeJYtj5O0JwnSsEBaTHVNmBDKglSQ-kaRSAsgRBAUcCk+UoC5WSVQgMAUHDF9hR1PV8UJY1TQ7LtrVtRMHXLZ1XTnGy7KVPkEG5c4hRwL9JCeQoRKBbRME9AVohiiBtE7CUxTLKA3FVJAQgLE0lUxVI7RpOtVWAJxKAkABaeTJVIflBWAGlElK6A4GYMMEDgF9uocpArLCj5Z1eLcvC6EBl13eS32PT9T0sH9bFIRwUBcB8n01Q9wMc5zzkoaDiWNRBiOQ8IojmjCOOIvCCMQ7Ropcy65viJIUkoUkHspKQyNeCipoBcy6NbRTmP49i9s4hieLh1iBN0qBsVxTy6RJOxyTm6yZMihH9uFFG+LR0mcqgdSnE07Tmz0tIDKQIyxtMwtaNBXAidsxVlVVJyXLSCo6VZrT4wkqRW0-es-uqxNTXDeikFliEwDASh0HufLEGYbtoggFAmXmQKdKZIdtcgUERrtE1mEwOA0CK6A5rgLoRsmqj+fsz7YoqeLwUSkdsYNNKMrd7LcrlgroGKsMyojHkquoYzzfqxqWray5OrvMberCSdBrVka0jGkrvd+GbAndhalqJFb3XfE8LHPaxfx2jiDtAl9joc4XzsuolRn+9JgdJIIlF+tAPuy77gd+2NAelsiSmMJ5Tmc1UzItZhLhuEFHggKID7uB5xQgUoudEaVWA4IQoCmK10CmKIplFcVLimAx2DIs+j6X2ZIqGgRIYikhIkAA)which shows that `JSON_Override.parse(str);` is not an error

If you can reproduce erroneous behaviour in TS Playground, feel free to send it.

**Comment by fatcerberus** [comment]
It absolutely *is* an error.  You changed the `let` to `const`.

**Comment by brunocrosier** [comment]
I used the `const` from your original comment

![image](https://user-images.githubusercontent.com/18399089/193177469-a599f507-f9ba-405f-bd14-b98d573a74d5.png)

I can see that there is an error with `let`

I will investigate that

**Comment by nopeless** [comment]
> If you know your string statically at compile time, why do you need to parse it anyway?

@Josh-Cena A case might be using `JSON.parse` to load large objects. Might be practical in browsers with low performance computers.

**Comment by nopeless** [comment]
@brunocrosier
https://262.ecma-international.org/6.0/#sec-json.parse
TL;DR: any JavaScript-like expression is a valid JSON

According to this, all of the values you provided should be parsable, including the empty string

**Comment by fatcerberus** [comment]
For the record, “this isn’t a useful change” and “this demonstrates a misunderstanding of how TypeScript works” are mere observations and don’t constitute CoC violations.  I think it’s fair to criticize a feature request on those grounds, no?

Anyway: types are not values, and they exist only at compile time.  The proposed checks for `JSON.parse` *only* work when the compiler has a way to know, at compile-time, *exactly* what is being passed to that function—because your proposed types, besides rejecting ill-formed JSON string literals, *also* disallow passing any variable typed as `string`.  That’s not a useful behavior and would make the function very painful to use for almost all real-world use cases.

**Comment by MartinJohns** [comment]
Your implementation of "JSON_Override" wrongly prevents leading or tailing whitespace characters.

**Comment by nopeless** [comment]
> Your implementation of "JSON_Override" wrongly prevents leading or tailing whitespace characters.
> 
> In general, as mentioned, this only works for string literals. A very narrow use case, but a whole buttload of added complexity, resulting in compilation slowdown and breaking changes (added global types). Especially such a generic name like `CircularReference` is a **really** bad idea.

I'm not sure why `CircularReference` is a bad idea. As long as the context is clear in the type file, it shouldn't matter at all

**Comment by MartinJohns** [comment]
It has such a generic name, odds are such a name is already in use, resulting in a breaking change. The introduction of such a utility type also means the team can't update it in the future (if a new feature would be better suited for such a type), without causing yet another breaking change (see `Omit<>`). And the team has the position to not add new utility types, unless they're required for declaration emit purposes. See https://github.com/microsoft/TypeScript/issues/39522#issuecomment-656201388:

> We've opted to not include utility type aliases in the lib unless they're required for declaration emit purposes.

**Comment by brunocrosier** [comment]
You are right. At the moment, the types in this PR at the moment are too strict - particularly because of `AllowedJSONParseText`.

I still believe there is a lot of benefit in disallowing the parsing of an empty string, or the stringification of a `bigint`

What about something like this? It disallows values we **know** cannot be parsed / stringified (as opposed to the previous solution which would try to codify the allowed ones as a type and resolve to `never` otherwise)

[TS Playground](https://www.typescriptlang.org/play?target=7#code/C4TwDgpgBAIglgZwIYBsUHsDuEAmApAZQHkA5ABSQCcEIAVCAD2CgF4oBydqAHygAMAJAG8EwSnAB2AcwC+w9gBoAul16KZqjgqgqeWqBpl8AUKEixEqDNnzESANVQBXaGwBGcKZOa94yNFi4hKQAQp4AkhLARG4AVhAAxsCm4NB+VoG2oRFRMfFJrFBCxlClUADaANYQIABcUKLi0noSTgC2bhCUegggHegoSvUeXlHGMnrFZRXVdQ1iklIt7Z3dvL39g-XpATbBJGFSkdFxickyxsbeXQBmSAnQ+wD6RABuXeI40FNlYFQ0AB5aFBGMAIBIcAh5k0pAA+AAUYKY9WBoPBkIs-msQTsFGodFBUAA-FAJBB3t0UdpKOS4BSifVEQALRD1JASEDaWb1RqLbSvZwQNkcgCUrFhUHZIBFwpAAG4SmVedI4DcQECEQKUC4USCmOioTtsVkHILiaTyV0oFSoDSwCh7l0GVBmazJRyuTUeQtpPzBbKxSwJVLtAg-g9ncqlrxWh0ujLoYsFdMo6r1bRNf6oKj9RDDZZdjjSI5tdASWSKdbs9SIPbHZRnfDY6sej6pCLykplmhQ+GIJG28s45QE1GFRdjF8Eg6aVAEugJKIoM83h84F96iuKZ8IJcAPR7qAAAWACAAtIxIEkL5RKOhKMYt2uvgA6P74+GcEX7w8n8+Xs4bzvB8n1vdcIDff4IHhAAicoAEZtAAJm0ABmbQABZtCUGCRQVS5jHnRdmAgNowFAAhBzYThjAPY9TwvBgr2AID70fOwXm3cDII-UjyJASiYTFOjQJ3HjARgmDYUuFAIGYWTgAAUTIiiqI4dh2NITjnwg98aHhBTlP4wTFmEw9RO4vSIABKMJSbdAQVve9v0Ihcl0aQp2CEKAYJudB0Bg+oYK1dd9hggx2AVCzXys+FGjw4wFOheDPO83z-MCnyQpNcKNCijjVzAmKoLisQEto38GIA68PjY6KINTNV4R+UpuSgJCJCgOiATPM8oBGbxxhcui-0Y5jWJAgquNfRqQGaxVSgpOYWumKAvlrFBloW1bpjJURcHqFadp2yRssO7bjsuqAtR1drOu63r+s8QaruOi5XtKd63u2i4ZBFIA)

**Comment by brunocrosier** [comment]
That's a good point re: the replacer. 

How about [this](https://www.typescriptlang.org/play?target=7#code/JYOwLgpgTgZghgYwgAgFIGUDyA5A+pgN2imABMUBvAKGVuQAc4oBnCAHgBVkIAPSEUs2TMwJEAHMAfAApIfAFzIuvfoOQBydcgA+yAAYASCiLHiAvkfUAaALpbd1s-Y1XkdnS+ROze5AH5kEAgiKGRFDlcoYOAQv0VZAAtgZkU4EABPVwBrCHTFE1BxVwI4ABsAVwhUjIBKZABeSWQ09Jrq9IBuGjoCiWAYdOkSiqrmjMiIelLEaDjkROT27Nz80ULissr2usaxzOFGJDne8Q8QcoBbACNoNuE1iS66e9N+weGtvYmpmag56XO12gHhONQA2jYzuVSqVXMxDhBjg9TrpATcoHcTl0zFQqOQENMosgEAB7EAiNBYPCEYhkUYYHD4EIkci4hnU5l0gB0Jze0nUFGQH1GACYQF5rMgITUqOymbTyDzkXzBcLFGKvK4wZpXOpmKR1DYZVQAPQm5AAATAzAAtLx6BAEGA7VAoCSoLKqfLXdzGCwIPz1MazZbrXaeA6nS63R65TSfYq-axpAAiMEARlcItcAGZXAAWVw2FM1Lq4qik8lgbgXehgdLoZENDTqU3mq22+2O53Ed2exnxlkQLlJgMQWv1xumOohuOcxNMVhsFMpyS40oQasbsAAUQnDab9Rb-Y5CuHo+k273dYP0+Qs69g99i-YJyaAJJ3Fd7uNlYpJmbAVkBTGASRJFNFBTYYyHZFMJS6OczxHF9pBMUsqG3F500AwUQLAiDgOg0hYPgk9vSHZD-VQ0R03Q3EgA)? 

- [x] No utility types 
- [x] Only updates `JSON.parse` to disallow arguments we know cannot be parsed

**Comment by MartinJohns** [comment]
I'm no member of the TypeScript team, so I have no say. But I don't see the point in vastly overcomplicating these types for a negligible use case. This would only work for inferred string literal types, and I can't think of a reason why you would pass a string literal to `JSON.parse()` instead of having the object literal directly.

Honestly, this would be much better suited as an ESLint rule.

**Comment by brunocrosier** [comment]
Yeah, you're right actually. My bad
